guacamole/src/main/java/org/apache/guacamole/rest/user/UserResource.java

```diff
@@ -145,9 +150,51 @@ public UserHistoryResource getUserHistory()
     @Override
     public void updateObject(APIUser modifiedObject) throws GuacamoleException {
 
-        // A user may not use this endpoint to modify himself
-        if (userContext.self().getIdentifier().equals(modifiedObject.getUsername()))
-            throw new GuacamoleSecurityException("Permission denied.");
+        User currentUser = userContext.self();
+
+        // A user may not use this endpoint to modify themself, except in the case
+        // that they are modifying one of the user attributes explicitly exposed
+        // in the user preferences form
+        if (currentUser.getIdentifier().equals(modifiedObject.getUsername())) {
+
+            // A user may not use this endpoint to update their password
+            if (currentUser.getPassword() != null)
+                throw new GuacamoleSecurityException(
+                        "Permission denied. The password update endpoint must"
+                        + " be used to change the current user's password.");
+
+            // All attributes exposed in the preferences forms
+            Set<String> preferenceAttributes = (
+                    userContext.getUserPreferenceAttributes().stream()
+                    .flatMap(form -> form.getFields().stream().map(
+                            field -> field.getName())))
+                    .collect(Collectors.toSet());
+
+            // Go through every attribute value and check if it's changed
+            Iterator<String> keyIterator = modifiedObject.getAttributes().keySet().iterator();
+            while(keyIterator.hasNext()) {
+
+                String key = keyIterator.next();
+                String newValue = modifiedObject.getAttributes().get(key);
+
+                // If it's not a preference attribute, editing is not allowed
+                if (!preferenceAttributes.contains(key)) {
+
+                    String currentValue = currentUser.getAttributes().get(key);
+
+                    // If the value of the attribute has been modified
+                    if (
+                            !(currentValue == null && newValue == null) && (
+                                (currentValue == null && newValue != null) ||
+                                !currentValue.equals(newValue)
+                            )
+                    )
+                        throw new GuacamoleSecurityException(
+                            "Permission denied. Only user preference attributes"
+                            + " can be modified for the current user.");
+                }
+            }
+        }
```

Comment:
This is already provided in principle via `filterAttributes()` on `DirectoryObjectTranslator`:

https://github.com/apache/guacamole-client/blob/754e9649f1fa0ba225ee42b56ded64bc283d17df/guacamole/src/main/java/org/apache/guacamole/rest/directory/DirectoryObjectTranslator.java#L119-L135

which is already invoked for every received user object for update:

https://github.com/apache/guacamole-client/blob/754e9649f1fa0ba225ee42b56ded64bc283d17df/guacamole/src/main/java/org/apache/guacamole/rest/user/UserObjectTranslator.java#L58-L66

To that end:

* I think we should leverage the existing `filterAttributes()`.
* Won't the existing `filterAttributes()` call filter out anything that isn't declared as a user attribute here?

I wonder if perhaps the existing call to `filterAttributes()` should (1) consider whether the user is attempting to modify themselves and (2) filter based on user preference attributes instead of user attributes in that case.

Reply:
Ah yes, that would be much simpler. I've updated it to leverage `filterAttributes()`. 

As to your question about the existing call filtering out the attributes - it doesn't, because the preference attributes for the KSM extension are also included in the regular user attributes.